expression/integration_test.go

```diff
@@ -7492,3 +7492,108 @@ func (s *testIntegrationSuite) TestIssue20180(c *C) {
 	tk.MustExec("insert into t values('b');")
 	tk.MustQuery("select * from t where a > 1  and a = \"b\";").Check(testkit.Rows("b"))
 }
+
+func (s *testIntegrationSuite) TestIssue19892(c *C) {
+	defer s.cleanEnv(c)
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("USE test")
+	tk.MustExec("CREATE TABLE dd(a date, b datetime)")
+
+	// check NO_ZERO_DATE
+	{
+		tk.MustExec("SET sql_mode=''")
+		{
+			tk.MustExec("TRUNCATE TABLE dd")
+			tk.MustExec("INSERT INTO dd(a) values('0000-00-00')")
+			tk.MustQuery("SHOW WARNINGS").Check(testkit.Rows())
+			tk.MustQuery("SELECT a FROM dd").Check(testkit.Rows("0000-00-00"))
+
+			tk.MustExec("TRUNCATE TABLE dd")
+			tk.MustExec("INSERT INTO dd(b) values('2000-10-01')")
+			tk.MustExec("UPDATE dd SET b = '0000-00-00'")
+			tk.MustQuery("SHOW WARNINGS").Check(testkit.Rows())
+			tk.MustQuery("SELECT b FROM dd").Check(testkit.Rows("0000-00-00 00:00:00"))
+		}
+
+		tk.MustExec("SET sql_mode='NO_ZERO_DATE'")
+		{
+			tk.MustExec("TRUNCATE TABLE dd")
+			tk.MustExec("INSERT INTO dd(b) values('0000-00-00')")
+			tk.MustQuery("SHOW WARNINGS").Check(testkit.Rows("Warning 1292 Incorrect datetime value: '0000-00-00 00:00:00'"))
+			tk.MustQuery("SELECT b FROM dd").Check(testkit.Rows("0000-00-00 00:00:00"))
+
+			tk.MustExec("TRUNCATE TABLE dd")
+			tk.MustExec("INSERT INTO dd(a) values('2000-10-01')")
+			tk.MustExec("UPDATE dd SET a = '0000-00-00'")
+			// tk.MustQuery("SHOW WARNINGS").Check(testkit.Rows("Warning 1292 Incorrect date value: '0000-00-00'"))
+			tk.MustQuery("SELECT a FROM dd").Check(testkit.Rows("0000-00-00"))
+		}
+
+		tk.MustExec("SET sql_mode='NO_ZERO_DATE,STRICT_TRANS_TABLES'")
+		{
+			tk.MustExec("TRUNCATE TABLE dd")
+			tk.MustGetErrMsg("INSERT INTO dd(b) VALUES ('0000-00-00')", "[table:1366]Incorrect datetime value: '0000-00-00' for column 'b' at row 1")
+			tk.MustExec("INSERT IGNORE INTO dd(b) VALUES ('0000-00-00')")
+			tk.MustQuery("SHOW WARNINGS").Check(testkit.Rows("Warning 1292 Incorrect datetime value: '0000-00-00 00:00:00'"))
+			tk.MustQuery("SELECT b FROM dd").Check(testkit.Rows("0000-00-00 00:00:00"))
+
+			tk.MustExec("TRUNCATE TABLE dd")
+			tk.MustExec("INSERT INTO dd(b) values('2000-10-01')")
+			tk.MustGetErrMsg("UPDATE dd SET b = '0000-00-00'", "[types:1292]Incorrect datetime value: '0000-00-00 00:00:00'")
+			tk.MustExec("UPDATE IGNORE dd SET b = '0000-00-00'")
+			// FIXME: has two warning rows
+			// tk.MustQuery("SHOW WARNINGS").Check(testkit.Rows("Warning 1292 Incorrect datetime value: '0000-00-00 00:00:00'"))
```

Comment:
The reason for the two warnings is that the executor calls `CastValue` twice during the `Update`. But this has nothing to do with your modification. You can use `tk.MustQuery("SHOW WARNINGS").Check(testkit.Rows("Warning 1292 Incorrect datetime value: '0000-00-00 00:00:00'", "Warning 1292 Incorrect datetime value: '0000-00-00 00:00:00'"))` here to indicate that there is indeed Two lines of errors, then create a new issue with two lines of errors.

Reply:
> The reason for the two warnings is that the executor calls CastValue twice during the Update

Thanks for your advice. In `UPDATE` execute, it does exist twice. 

eg: the exec trace of `UPDATE dd SET b = '0000-00-00'` as below:

1: cast from `types.KindString` to `mysql.TypeDatetime`;
2. cast from `types.KindMysqlTime` to `mysql.TypeDatetime`;

```bash
# the first CastValue
# cast from `types.KindString` to `mysql.TypeDatetime`

 0  0x0000000001ff4afb in github.com/pingcap/tidb/table.CastValue
    at ./table/column.go:168
 1  0x0000000002aaba49 in github.com/pingcap/tidb/executor.(*UpdateExec).fastComposeNewRow
    at ./executor/update.go:232
 2  0x0000000002acf1c6 in github.com/pingcap/tidb/executor.(*UpdateExec).fastComposeNewRow-fm
    at ./executor/update.go:215
 3  0x0000000002aaab6a in github.com/pingcap/tidb/executor.(*UpdateExec).updateRows
    at ./executor/update.go:185
 4  0x0000000002aaa74c in github.com/pingcap/tidb/executor.(*UpdateExec).Next
    at ./executor/update.go:134
    ......

# the second CastValue
# cast from `types.KindMysqlTime` to `mysql.TypeDatetime`;

(dlv) bt
 0  0x0000000001ff4afb in github.com/pingcap/tidb/table.CastValue
    at ./table/column.go:168
 1  0x0000000002ab0243 in github.com/pingcap/tidb/executor.updateRecord
    at ./executor/write.go:77
 2  0x0000000002aaa30a in github.com/pingcap/tidb/executor.(*UpdateExec).exec
    at ./executor/update.go:104
 3  0x0000000002aaac51 in github.com/pingcap/tidb/executor.(*UpdateExec).updateRows
    at ./executor/update.go:189
 4  0x0000000002aaa74c in github.com/pingcap/tidb/executor.(*UpdateExec).Next
    at ./executor/update.go:134

 5  0x00000000029c3039 in github.com/pingcap/tidb/executor.Next
    at ./executor/executor.go:268
```

IMO, we can ignore the warning message in the second cast operation.